### DISCUSSION TITLE
EOS/EOB NALUs option

### DISCUSSION THREAD

**Discussion by Nico83500** [quietvoid/dovi_tool#288] [General]
Hi,
I would like to know the impact of the --eos-before-el option ?
Thank you.

**Comment by quietvoid** [comment]
It does what it says:
> Write the EOS/EOB NALUs before the EL. Defaults to false.
This flag enables the same behaviour yusesope's mux script.
Enabling this therefore results in identical output using dovi_tool.

> In recent versions of MakeMKV, EOS/EOB NALUs are removed.

EOS/EOB NALUs usually signal the end of a bitstream OR segment (in the case of seamless branching).
So when muxing the EL stream into a BL containing EOS/EOB, the EL is inserted either before or after depending on the setting.

By default the EL is inserted before EOS/EOB from the BL, to avoid issues with parsers stopping early when encountering one and ignoring the EL.

It should have no impact since they are not commonly in the middle of the bitstream.
The option is only there to reproduce bitstreams the same as yusesope's script.

In the HEVC specification, EOS/EOB type should always be the last NALUs in the access unit.
This is how `inject-rpu` behaves as well, inserting the UNSPEC62 (RPU) NALU before EOS/EOB.

FFmpeg is the only known parser that may ignore the RPU before EOS/EOB, because it always expects that the AU ends with UNSPEC62.

**Comment by Nico83500** [comment]
Thank you for this explanation !
So the default option (false) seems to be the best choice. I don't know why to use yusesope's script bitstreams, in which case it could be helpful ?
mkvtoolnix seems to do like this (false) but DGDemux use --eso-before-el

**Reply by quietvoid** [reply]
yusesope's script was the original method to the streams interleaving, it's not really relevant today.

**Reply by Nico83500** [reply]
Thank you !
